### PR TITLE
OTC-114: comparing reconciliation data and update payment status

### PR DESCRIPTION
**Pull request by sniedzielski**
TICKET: https://openimis.atlassian.net/browse/OTC-114

Part related to make update of reconciliated payment from GePG. 
- update PaymentStatus = 5 for payment status < 5 and expectedAmount(IMIS)=amountPaid(GePG) and ReceivedAmount(IMIS) =amountPaid(GePG).
- update PaymentStatus = -5 for payment which are paid in GePG, but no info about paid in IMIS.

### CODE REVIEW

File: OpenImis.ePayment/Escape/Payment/Controllers/PaymentController.cs [hunk 1]
```diff
@@ -117,6 +117,29 @@ public IActionResult GetReconciliation([FromBody] gepgSpReconcResp model)
                 var gepgFile = new GepgFoldersCreating("Reconc", "Data", reconc, env);
                 gepgFile.putToTargetFolderPayment();
 
+                foreach (var recon in model.ReconcTrxInf) 
+                {
+                    var paymentToCompare = imisPayment.GetPaymentToReconciliate(new { billId = recon.SpBillId, paidAmount = recon.PaidAmt });
+                    if (paymentToCompare != null)
+                    {
+                        int billId = (int) paymentToCompare.GetType().GetProperty("paymentStatus").GetValue(paymentToCompare);
+                        if (billId < 5)
```
Comment (dragos-dobre) [inline OpenImis.ePayment/Escape/Payment/Controllers/PaymentController.cs:126]: Please create a PaymentStatus static list with all statuses as constants and use them in the code for better understanding. As for GepgCodeResponses.

File: OpenImis.ePayment/Data/ImisBasePayment.cs [hunk 2]
```diff
@@ -821,6 +821,117 @@ public void CancelPayment(int payment_id)
...
+        public object GetPaymentToReconciliate(object payment)
+        {
+            object result = null;
+            double paidAmount = Convert.ToDouble(payment.GetType().GetProperty("paidAmount").GetValue(payment));
+            SqlParameter[] parameters = {
+                new SqlParameter("@Id", payment.GetType().GetProperty("billId").GetValue(payment)),
+                new SqlParameter("@paidAmount", paidAmount),
+            };
+           
+            var sSQL = @"SELECT PaymentId, ExpectedAmount, ReceivedAmount, PaymentStatus FROM tblPayment WHERE PaymentId=@Id And PaymentStatus<=5 And ExpectedAmount=@paidAmount And ValidityTo is Null";
```
Comment (dragos-dobre) [inline OpenImis.ePayment/Data/ImisBasePayment.cs:833]: Can you use some Declaration in the SQL or add the PaymentStatus as variable and use the constants defined based on a later comment in this PR?

File: OpenImis.ePayment/Data/ImisBasePayment.cs [hunk 3]
```diff
@@ -821,6 +821,117 @@ public void CancelPayment(int payment_id)
             }
         }
 
+        public object GetPaymentToReconciliate(object payment)
```
Comment (dragos-dobre) [inline OpenImis.ePayment/Data/ImisBasePayment.cs:824]: Why object type and not an actual type?

File: OpenImis.ePayment/Escape/Payment/Controllers/PaymentController.cs (hunk 1 above)
Comment (dragos-dobre) [inline OpenImis.ePayment/Escape/Payment/Controllers/PaymentController.cs:126]: I think this is a comparation between a PaymentId and a PaymentStatus. Is this correct?

**Review by dragos-dobre — CHANGES_REQUESTED** [review]

File: OpenImis.ePayment/Escape/Payment/Controllers/PaymentController.cs (hunk 1 above)
Comment (sniedzielski) [reply OpenImis.ePayment/Escape/Payment/Controllers/PaymentController.cs:126]: Yes. You are right. My gap. I fix this!

Comment (sniedzielski) [reply OpenImis.ePayment/Escape/Payment/Controllers/PaymentController.cs:126]: ok!

File: OpenImis.ePayment/Data/ImisBasePayment.cs (hunk 3 above)
Comment (sniedzielski) [reply OpenImis.ePayment/Data/ImisBasePayment.cs:824]: Ok. You are right. I wanted to use generic object. But using current Reconc type for sure might be more readable in code.  I change this.

File: OpenImis.ePayment/Data/ImisBasePayment.cs (hunk 2 above)
Comment (sniedzielski) [reply OpenImis.ePayment/Data/ImisBasePayment.cs:833]: Ok. Sure! Clear for me what do you want here.

File: OpenImis.ePayment/Data/ImisBasePayment.cs [hunk 4]
```diff
@@ -803,11 +803,12 @@ public List<ReconciliationItem> ProvideReconciliationData(ReconciliationRequest
         public void CancelPayment(int payment_id)
         {
             var sSQL = @"UPDATE tblPayment
-                         SET PaymentStatus = 0
+                         SET PaymentStatus = @PaymentStatus
```
Comment (sniedzielski) [inline OpenImis.ePayment/Data/ImisBasePayment.cs:806]: @dragos-dobre  btw I changed here in the cancellation to take from the static class values according to your tip from review.

File: OpenImis.ePayment/Data/ImisBasePayment.cs [hunk 5]
```diff
@@ -803,11 +803,12 @@ public List<ReconciliationItem> ProvideReconciliationData(ReconciliationRequest
...
                          WHERE PaymentID = @PaymentID;";
 
             SqlParameter[] parameters = {
-                new SqlParameter("@PaymentID", payment_id)
+                new SqlParameter("@PaymentID", payment_id),
+                new SqlParameter("@PaymentStatus", PaymentStatus.Cancelled)
```
Comment (sniedzielski) [inline OpenImis.ePayment/Data/ImisBasePayment.cs:811]: @dragos-dobre btw I changed here in the cancellation to take from the static class values according to your tip from review.

File: OpenImis.ePayment/Escape/Payment/Controllers/PaymentController.cs (hunk 1 above)
Comment (sniedzielski) [reply OpenImis.ePayment/Escape/Payment/Controllers/PaymentController.cs]: @dragos-dobre I created static class in similar way as for GePGCodeResponses. It should be the same like having static List. Then we have in code clear which payment status we are using.

**Comment by sniedzielski** [comment]
@dragos-dobre  You can recheck code after the review! I made some changes. Instead of static list of 'payment status' I used static class in simillar way as I did for GePGCodeResponses. Then in code we can see which status is taken/updated etc in much more readable way.

File: OpenImis.ePayment/Escape/Payment/Models/PaymentStatus.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public static class PaymentStatus
+    {
+        public static int FailedReconciliated  = -5;
+        public static int FailedReceived = -4;
+        public static int FailedControlNumberReceived = -3;
+        public static int FailedSent = -2;
+        public static int FailedRequested = -1;
+        public static int Cancelled = 0;
+        public static int Requested = 1;
```
Comment (dragos-dobre) [inline OpenImis.ePayment/Escape/Payment/Models/PaymentStatus.cs:12]: Status 1 is Payment created. Please check.

File: OpenImis.ePayment/Escape/Payment/Models/PaymentStatus.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+        public static int Sent = 2;
```
Comment (dragos-dobre) [inline OpenImis.ePayment/Escape/Payment/Models/PaymentStatus.cs:13]: Status 2 is ControlNumberRequested.

**Review by dragos-dobre — CHANGES_REQUESTED** [review]

File: OpenImis.ePayment/Escape/Payment/Models/PaymentStatus.cs (hunk 7 above)
Comment (sniedzielski) [reply OpenImis.ePayment/Escape/Payment/Models/PaymentStatus.cs:13]: Ok. I change this.

File: OpenImis.ePayment/Escape/Payment/Models/PaymentStatus.cs (hunk 6 above)
Comment (sniedzielski) [reply OpenImis.ePayment/Escape/Payment/Models/PaymentStatus.cs:12]: Ok. I change this.

**Comment by sniedzielski** [comment]
@dragos-dobre I updated payment status names (for 1 and 2)

**Review by dragos-dobre — APPROVED** [review]

**Pull request merged**